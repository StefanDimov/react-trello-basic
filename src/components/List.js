import React from 'react'
import Card from './Card'
import { Panel, ListGroup, ListGroupItem } from 'react-bootstrap'

export default class List extends React.Component {
    render() {
        const {title, cards, onCardClick, children} = this.props

        const cardElements = cards.map(card =>
            <ListGroupItem onClick={onCardClick} key={card.title}>
                <Card title={card.title}/>
            </ListGroupItem>
        )

        return (
            <Panel header={title}>
                {(cardElements.length > 0) && <ListGroup>{cardElements}</ListGroup>}
                {children}
            </Panel>
        )
    }
}

const { string, array, func } = React.PropTypes
List.propTypes = {
    title: string.isRequired,
    cards: array.isRequired,
    onCardClick: func.isRequired
}
