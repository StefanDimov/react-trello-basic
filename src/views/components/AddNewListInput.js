import React from 'react'
import { FormGroup, FormControl } from 'react-bootstrap'

export default class AddNewListInput extends React.Component {
    constructor(props) {
        super(props)

        this.state = { value: '' }

        this.changeValueHandler = this.changeValueHandler.bind(this)
        this.onSubmit = this.onSubmit.bind(this)
    }

    changeValueHandler(e) {
        this.setState({ value: e.target.value })
    }

    onSubmit(e) {
        e.preventDefault();
        this.props.onCreateList(this.state.value)
        this.setState({ value: '' })
    }

    render() {
        return (
            <form className="my-list-container-item" onSubmit={this.onSubmit}>
                <FormGroup>
                    <FormControl
                        type="text"
                        value={this.state.value}
                        onChange={this.changeValueHandler}
                        placeholder="Add new list..." />
                </FormGroup>
            </form>
        )
    }
}

const { func } = React.PropTypes
AddNewListInput.propTypes = {
    onCreateList: func.isRequired
}
