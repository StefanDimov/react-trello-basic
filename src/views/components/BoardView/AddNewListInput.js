import React from 'react'
import { FormGroup, FormControl } from 'react-bootstrap'

/**
 * Component with an input for adding a new list. Meant to be used in Board Component.
 * @class comp/AddNewListInput
 * @param {function} onCreateList callback for input submit
 */
export default class AddNewListInput extends React.Component {
    constructor(props) { // eslint-disable-line
        super(props)

        this.state = { value: '' }

        this.changeValueHandler = this.changeValueHandler.bind(this)
        this.onSubmit = this.onSubmit.bind(this)
    }

    /**
     * Controls the input change event
     * @private
     * @param {object} e event object
     */
    changeValueHandler(e) {
        this.setState({ value: e.target.value })
    }

    /**
     *  Controls the form submit event
     *  @private
     *  @param {object} e event object
     */
    onSubmit(e) {
        e.preventDefault()
        this.props.onCreateList(this.state.value)
        this.setState({ value: '' })
    }

    render() { // eslint-disable-line
        return (
            <form className="my-list-container-item" onSubmit={this.onSubmit}>
                <FormGroup>
                    <FormControl
                        type="text"
                        value={this.state.value}
                        onChange={this.changeValueHandler}
                        placeholder="Add new list..." />
                </FormGroup>
            </form>
        )
    }
}

const { func } = React.PropTypes
AddNewListInput.propTypes = {
    onCreateList: func.isRequired
}
