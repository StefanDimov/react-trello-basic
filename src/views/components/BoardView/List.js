import React from 'react'
import Card from './Card'
import { Panel, ListGroup, ListGroupItem } from 'react-bootstrap'

import listUtils from '../../../utils/listUtils'

export default class List extends React.Component {
    render() {
        const { list, onCardClick, children} = this.props
        const { title, cards } = list

        const cardElements = cards.map(card =>
            <ListGroupItem onClick={() => onCardClick(card)} key={card.id}>
                <Card card={card}/>
            </ListGroupItem>
        )

        return (
            <Panel header={title} className="my-list-container-item">
                {(cardElements.length > 0) && <ListGroup>{cardElements}</ListGroup>}
                {children}
            </Panel>
        )
    }
}

List.propTypes = {
    list: listUtils.listPropType.isRequired,
    onCardClick: React.PropTypes.func.isRequired
}
