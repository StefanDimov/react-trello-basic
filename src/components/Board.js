import React from 'react'
import List from './List'
import { Navbar, Grid, Row, Col, Button } from 'react-bootstrap'

export default class Board extends React.Component {
    render() {
        const { title, lists, onCardClick, onCreateCardClick } = this.props

        const listElements = lists.map(list =>
            <Col key={list.title} sm={3}>
                <List title={list.title} cards={list.cards} onCardClick={onCardClick}>
                    <Button bsStyle="primary" block onClick={() => onCreateCardClick(list.id)}>Add Card</Button>
                </List>
            </Col>
        )

        return (
            <div>
                <Navbar fluid={true}>
                    <Navbar.Header>
                        <Navbar.Brand>{title}</Navbar.Brand>
                    </Navbar.Header>
                </Navbar>

                <Grid fluid={true}>
                    <Row>{listElements}</Row>
                </Grid>
            </div>
        )
    }
}

const { string, array, func } = React.PropTypes
Board.propTypes = {
    title: string.isRequired,
    lists: array.isRequired,
    onCardClick: func.isRequired,
    onCreateCardClick: func.isRequired
}
