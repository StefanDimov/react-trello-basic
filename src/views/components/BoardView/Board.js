import React from 'react'
import { Navbar, Nav, NavItem, Button, Glyphicon } from 'react-bootstrap'

import * as boardUtils from '../../../utils/boardUtils'
import CardList from './CardList'
import ListsWrapper from './ListsWrapper'
import AddNewListInput from './AddNewListInput'

/**
 * Component that visualizes a Board's lists and holds all the functionality regarding them.
 * @class comp/Board
 * @param {object} board The Board to be visualized
 * @param {function} onCardClick callback for when a card is clicked
 * @param {function} onCreateCard callback to which the data for a new card will be passed
 * @param {function} onCreateNewList callback to which the data for a new list will be passed
 * @param {function} onSaveBoard callback to which the data for updating a board will be passed
 */
export default class Board extends React.Component {
    render() { // eslint-disable-line
        const { board, onCardClick, onCreateCard, onCreateNewList, onSaveBoard } = this.props
        const { title, lists } = board

        const listElements = lists.map(list =>
            <CardList key={list.id} list={list} onCardClick={onCardClick}>
                <Button bsStyle="primary" block onClick={() => onCreateCard(list.id)}>Add Card</Button>
            </CardList>
        )

        return (
            <div className="my-navbar-top-padding">
                <Navbar fluid={true} fixedTop={true}>
                    <Navbar.Header>
                        <Navbar.Brand>{title}</Navbar.Brand>
                    </Navbar.Header>
                    <Nav pullRight={true}>
                        <NavItem onClick={() => { onSaveBoard(board) }}>
                            <Glyphicon glyph="floppy-disk" /> Save Board
                        </NavItem>
                        <NavItem href="/#">
                            <Glyphicon glyph="th-large" /> Boards
                        </NavItem>
                    </Nav>
                </Navbar>

                <ListsWrapper>
                    {listElements}
                    <AddNewListInput onCreateList={onCreateNewList} />
                </ListsWrapper>
            </div>
        )
    }
}

const { func } = React.PropTypes
Board.propTypes = {
    board: boardUtils.boardPropType.isRequired,
    onCardClick: func.isRequired,
    onCreateCard: func.isRequired,
    onCreateNewList: func.isRequired,
    onSaveBoard: func.isRequired
}
