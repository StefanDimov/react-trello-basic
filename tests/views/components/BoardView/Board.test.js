import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Board from '../../../../src/views/components/BoardView/Board'
import List from '../../../../src/views/components/BoardView/List'
import AddNewListInput from '../../../../src/views/components/BoardView/AddNewListInput'
import { Button } from 'react-bootstrap'

import { getEmptyBoard, getBoardWithLists } from '../../../_mocks/Board.mocks'

describe('Board', () => {

    const onCardClick = jest.fn()
    const onCreateCardClick = jest.fn()
    const onCreateNewList = jest.fn()

    let boardWithTitle, boardWithTitleAndOneList, boardWithTitleAndTwoLists

    beforeEach(() => {
        boardWithTitle = getEmptyBoard()
        boardWithTitleAndOneList = getBoardWithLists(undefined, 1)
        boardWithTitleAndTwoLists = getBoardWithLists(undefined, 2)

        onCardClick.mockClear()
        onCreateCardClick.mockClear()
        onCreateNewList.mockClear()
    })

    describe('rendering', () => {
        it('should render properly with a title and no lists', () => {
            const tree = renderer.create(
                <Board
                    board={boardWithTitle}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })

        it('should render properly with a title and a single list', () => {
            const tree = renderer.create(
                <Board
                    board={boardWithTitleAndOneList}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })

        it('should render properly with a title and two lists', () => {
            const tree = renderer.create(
                <Board
                    board={boardWithTitleAndTwoLists}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })
    })

    describe('lists wrapper', () => {
        it('should have a AddNewListInput child with onCreateNewList passed to it', () => {
            const wrapper = shallow(
                <Board
                    board={boardWithTitleAndOneList}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            )

            expect(wrapper.find(AddNewListInput).prop('onCreateList')).toBe(onCreateNewList)
        })
    })

    describe('lists', () => {
        it('should be passed onCardClick func', () => {
            const wrapper = shallow(
                <Board
                    board={boardWithTitleAndOneList}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            )

            expect(wrapper.find(List).prop('onCardClick')).toBe(onCardClick)
        })

        it('should have a button that calls onCreateCardClick with list id', () => {
            const wrapper = shallow(
                <Board
                    board={boardWithTitleAndOneList}
                    onCardClick={onCardClick}
                    onCreateCardClick={onCreateCardClick}
                    onCreateNewList={onCreateNewList} />
            )

            wrapper.find(List).find(Button).simulate('click')

            expect(onCreateCardClick).toHaveBeenCalledTimes(1)
            expect(onCreateCardClick).toHaveBeenCalledWith(boardWithTitleAndOneList.lists[0].id)
        })
    })
})
