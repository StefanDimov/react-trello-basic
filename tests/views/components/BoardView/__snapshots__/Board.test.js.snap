exports[`Board rendering should render properly with a title and a single list 1`] = `
<div
  className="my-navbar-top-padding">
  <nav
    className="navbar navbar-default navbar-fixed-top">
    <div
      className="container-fluid">
      <div
        className="navbar-header">
        <span
          className="navbar-brand">
          boardTitle
        </span>
      </div>
      <ul
        className="nav navbar-nav navbar-right"
        role={null}>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="#"
            onClick={[Function]}
            role="button">
            <span
              className="glyphicon glyphicon-floppy-disk" />
             Save Board
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="/#"
            onClick={[Function]}>
            <span
              className="glyphicon glyphicon-th-large" />
             Boards
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="my-lists-container">
    <div
      className="my-list-container-item panel panel-default"
      id={undefined}>
      <div
        className="panel-heading">
        list title0
      </div>
      <div
        className="panel-body">
        <button
          className="btn btn-primary btn-block"
          disabled={false}
          onClick={[Function]}
          type="button">
          Add Card
        </button>
      </div>
    </div>
    <form
      className="my-list-container-item"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new list..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;

exports[`Board rendering should render properly with a title and no lists 1`] = `
<div
  className="my-navbar-top-padding">
  <nav
    className="navbar navbar-default navbar-fixed-top">
    <div
      className="container-fluid">
      <div
        className="navbar-header">
        <span
          className="navbar-brand">
          boardTitle
        </span>
      </div>
      <ul
        className="nav navbar-nav navbar-right"
        role={null}>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="#"
            onClick={[Function]}
            role="button">
            <span
              className="glyphicon glyphicon-floppy-disk" />
             Save Board
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="/#"
            onClick={[Function]}>
            <span
              className="glyphicon glyphicon-th-large" />
             Boards
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="my-lists-container">
    <form
      className="my-list-container-item"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new list..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;

exports[`Board rendering should render properly with a title and two lists 1`] = `
<div
  className="my-navbar-top-padding">
  <nav
    className="navbar navbar-default navbar-fixed-top">
    <div
      className="container-fluid">
      <div
        className="navbar-header">
        <span
          className="navbar-brand">
          boardTitle
        </span>
      </div>
      <ul
        className="nav navbar-nav navbar-right"
        role={null}>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="#"
            onClick={[Function]}
            role="button">
            <span
              className="glyphicon glyphicon-floppy-disk" />
             Save Board
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}>
          <a
            href="/#"
            onClick={[Function]}>
            <span
              className="glyphicon glyphicon-th-large" />
             Boards
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="my-lists-container">
    <div
      className="my-list-container-item panel panel-default"
      id={undefined}>
      <div
        className="panel-heading">
        list title0
      </div>
      <div
        className="panel-body">
        <button
          className="btn btn-primary btn-block"
          disabled={false}
          onClick={[Function]}
          type="button">
          Add Card
        </button>
      </div>
    </div>
    <div
      className="my-list-container-item panel panel-default"
      id={undefined}>
      <div
        className="panel-heading">
        list title1
      </div>
      <div
        className="panel-body">
        <button
          className="btn btn-primary btn-block"
          disabled={false}
          onClick={[Function]}
          type="button">
          Add Card
        </button>
      </div>
    </div>
    <form
      className="my-list-container-item"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new list..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;
